index.js

```diff
@@ -0,0 +1,745 @@
+        }
+      };
+    };
+
+    var satisfactoryTypes =
+    function(name, constraints, $typeVarMap, expType, value, index) {
+      return _satisfactoryTypes(name, constraints, $typeVarMap, value, index)
+                               (expType, [value], id);
+    };
+
+    var curry = function(name, constraints, expArgTypes, expRetType,
+                         _typeVarMap, _values, _indexes, impl) {
+      return arity(_indexes.length, function() {
+        var delta = _indexes.length - arguments.length;
+        if (delta < 0) {
+          throw new TypeError(
+            LEFT_SINGLE_QUOTATION_MARK + name + RIGHT_SINGLE_QUOTATION_MARK +
+            ' requires ' + numArgs(expArgTypes.length) + ';' +
+            ' received ' + numArgs(expArgTypes.length - delta)
+          );
+        }
+        var $typeVarMap = {};
+        for (var typeVarName in _typeVarMap) {
+          $typeVarMap[typeVarName] = _typeVarMap[typeVarName].slice();
+        }
+        var values = _values.slice();
+        var indexes = [];
+        for (var idx = 0; idx < _indexes.length; idx += 1) {
+          var index = _indexes[idx];
+
+          if (idx < arguments.length &&
+              !(typeof arguments[idx] === 'object' &&
+                arguments[idx] != null &&
+                arguments[idx]['@@functional/placeholder'] === true)) {
+
+            var value = arguments[idx];
+            var expType = expArgTypes[index];
+            if (!expType.test(value) ||
+                isEmpty(satisfactoryTypes(name, constraints, $typeVarMap,
+                                          expType, value, index))) {
+              throw invalidArgument(name,
+                                    replaceTypeVars($typeVarMap)(expType),
+                                    value,
+                                    index);
+            }
+            values[index] = value;
+          } else {
+            indexes.push(index);
+          }
+        }
+        if (isEmpty(indexes)) {
+          var returnValue = impl.apply(this, values);
+          if (!expRetType.test(returnValue)) {
+            throw invalidReturnValue(name, [expRetType], returnValue);
+          }
+          satisfactoryTypes(name, constraints, $typeVarMap,
+                            expRetType, returnValue, NaN);
+          return returnValue;
+        } else {
+          return curry(name, constraints, expArgTypes, expRetType,
+                       $typeVarMap, values, indexes, impl);
+        }
+      });
+    };
+
+    return function(name, constraints, expTypes, impl) {
+      var expArgTypes = expTypes.slice(0, -1);
+      var arity = expArgTypes.length;
+      if (arity > 9) {
+        throw new RangeError(
+          LEFT_SINGLE_QUOTATION_MARK + 'def' + RIGHT_SINGLE_QUOTATION_MARK +
+          ' cannot define a function with arity greater than nine'
+        );
+      }
+
+      assertExpectedTypesInEnvironment(name)(expTypes);
```

Comment:
On Friday, @albrow raised an excellent point while reviewing the README. He was surprised to discover that it's possible to use a type in a function definition *without* first including that type in the environment. This is because the environment is only important when type variables are involved. When one applies `a -> a -> a` to `42`, sanctuary-def must determine the types of which `42` is a member. This is done by filtering the list of types provided to `$.create`.

Alex and I agreed that it's confusing to require types to be included in the environment in just *some* cases. Rather than document the actual behaviour, we decided it's preferable to state that `def` only accepts types in its environment. We then decided to go a step further and make it a run-time error to reference a "foreign" type in a function definition.